contrib/rabbitmq/pkg/adapter/adapter.go

```diff
@@ -0,0 +1,237 @@
+
+	if err := a.initClient(); err != nil {
+		logger.Error("Failed to create cloudevent client", zap.Error(err))
+		return err
+	}
+
+	logger.Info("Starting with config: ", zap.Any("adapter", a))
+
+	conn, err := amqp.Dial(a.Brokers)
+	if err != nil {
+		logger.Error(err)
+	}
+	defer conn.Close()
+
+	ch, err := conn.Channel()
+	if err != nil {
+		logger.Error(err)
+	}
+	defer ch.Close()
+
+	queue, err := a.StartAmqpClient(ctx, &ch)
+
+	return a.pollForMessages(ctx, &ch, queue, stopCh)
+}
+
+func (a *Adapter) StartAmqpClient(ctx context.Context, ch *wabbit.Channel) (*wabbit.Queue, error) {
+	logger := logging.FromContext(ctx)
+	exchangeConfig := fillDefaultValuesForExchangeConfig(&a.ExchangeConfig, a.Topic)
+
+	err := (*ch).ExchangeDeclare(
+		exchangeConfig.Name,
+		exchangeConfig.TypeOf,
+		wabbit.Option{
+			"durable":  exchangeConfig.Durable,
+			"delete":   exchangeConfig.AutoDeleted,
+			"internal": exchangeConfig.Internal,
+			"noWait":   exchangeConfig.NoWait,
+		},)
+	if err != nil {
+		logger.Error(err)
+		return nil, err
+	}
+
+	queue, err := (*ch).QueueDeclare(
+		a.QueueConfig.Name,
+		wabbit.Option{
+			"durable":   a.QueueConfig.Durable,
+			"delete":    a.QueueConfig.DeleteWhenUnused,
+			"exclusive": a.QueueConfig.Exclusive,
+			"noWait":    a.QueueConfig.NoWait,
+		},)
+	if err != nil {
+		logger.Error(err)
+		return nil, err
+	}
+
+	routingKeys := strings.Split(a.QueueConfig.RoutingKey, ",")
+
+	for _, routingKey := range routingKeys {
+		err = (*ch).QueueBind(
+			queue.Name(),
+			routingKey,
+			exchangeConfig.Name,
+			wabbit.Option{
+				"noWait": a.QueueConfig.NoWait,
+			},)
+		if err != nil {
+			logger.Error(err)
+			return nil, err
+		}
+	}
+	return &queue, nil
+}
+
+func (a *Adapter) pollForMessages(ctx context.Context, channel *wabbit.Channel,
+	queue *wabbit.Queue, stopCh <-chan struct{}) error {
+	logger := logging.FromContext(ctx)
+
+	msgs, err := (*channel).Consume(
+		(*queue).Name(),
+		"",
+		wabbit.Option{
+			"noAck":     false,
+			"exclusive": a.QueueConfig.Exclusive,
+			"noLocal":   false,
+			"noWait":    a.QueueConfig.NoWait,
+		},)
+
+	if err != nil {
+		logger.Error(err)
+	}
+
+	for {
+		select {
+		case msg, ok := <-msgs:
+			if ok {
+				logger.Info("Received: ", zap.Any("value", string(msg.Body())))
+
+				if err := a.postMessage(ctx, msg); err != nil {
+					logger.Info("Error posting message: ", zap.Error(err))
+				}
+			} else {
+				return nil
+			}
+		case <-stopCh:
+			logger.Info("Shutting down...")
+			return nil
+		}
+	}
+}
+
+func (a *Adapter) postMessage(ctx context.Context, msg wabbit.Delivery) error {
+	logger := logging.FromContext(ctx)
+
+	extensions := map[string]interface{}{
+		"key": string(msg.MessageId()),
+	}
+	event := cloudevents.Event{
+		Context: cloudevents.EventContextV02{
+			SpecVersion: cloudevents.CloudEventsVersionV02,
+			Type:        eventType,
+			ID:          msg.MessageId(),
+//			Time:        &types.Timestamp{msg.Timestamp},
```

Comment:
Any reason you're not including a timestamp?

Reply:
Hi @duglin, this change required a minor enhancement to one of the libraries which I made. Timestamp is included now.